File: creational/typescript/abstract_factory_structured/interfaces.ts
export interface Button {
  click(): void;
  render(): void;
  getName(): string;
}

export interface Checkbox {
  check(): void;
  render(): void;
  getName(): string;
}

export interface GUIFactory {
  createButton(): Button;
  createCheckbox(): Checkbox;
}

export interface IDatabaseConnection {
  connect(): void;
  disconnect(): void;
  executeQuery(query: string): void;
}

export interface DatabaseCommand {
  execute(): void;
}

export interface DatabaseFactory {
  createConnection(): IDatabaseConnection;
  createCommand(cmd: string): DatabaseCommand;
}

File: creational/dart/abstract_factory_structured/interfaces.dart
abstract class Button {
  void click();
  void render();
  String get name;
}

abstract class Checkbox {
  void check();
  void render();
  String get name;
}

abstract class GUIFactory {
  Button createButton();
  Checkbox createCheckbox();
}

abstract class IDatabaseConnection {
  void connect();
  void disconnect();
  void executeQuery(String query);
}

abstract class DatabaseCommand {
  void execute();
}

abstract class DatabaseFactory {
  IDatabaseConnection createConnection();
  DatabaseCommand createCommand(String cmd);
}
